File: src/dashboard/apigateway/apigateway/common/tenant/request.py
#
# TencentBlueKing is pleased to support the open source community by making
# 蓝鲸智云 - API 网关(BlueKing - APIGateway) available.
# Copyright (C) 2025 Tencent. All rights reserved.
# Licensed under the MIT License (the "License"); you may not use this file except
# in compliance with the License. You may obtain a copy of the License at
#
#     http://opensource.org/licenses/MIT
#
# Unless required by applicable law or agreed to in writing, software distributed under
# the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
# either express or implied. See the License for the specific language governing permissions and
# limitations under the License.
#
# We undertake not to change the open source license (MIT license) applicable
# to the current version of the project delivered to anyone in the future.
#
from typing import Dict

from django.conf import settings

from .constants import (
    TENANT_ID_OPERATION,
    TENANT_MODE_SINGLE_DEFAULT_TENANT_ID,
    TenantModeEnum,
)


def get_user_tenant_id(request) -> str:
    if settings.ENABLE_MULTI_TENANT_MODE:
        return request.user.tenant_id
    return TENANT_MODE_SINGLE_DEFAULT_TENANT_ID


def gen_operation_tenant_header() -> Dict[str, str]:
    return {"X-Bk-Tenant-Id": TENANT_ID_OPERATION}


def gen_tenant_header(tenant_id: str) -> Dict[str, str]:
    return {"X-Bk-Tenant-Id": tenant_id}


def get_tenant_id_for_gateway_maintainers(gateway_tenant_mode: str, gateway_tenant_id: str) -> str:
    # 不开启多租户，返回 default
    if not settings.ENABLE_MULTI_TENANT_MODE:
        return TENANT_MODE_SINGLE_DEFAULT_TENANT_ID

    # if gateway_tenant_mode is global, return operation tenant id system
    if gateway_tenant_mode == TenantModeEnum.GLOBAL.value:
        return TENANT_ID_OPERATION

    return gateway_tenant_id

File: src/mcp-proxy/pkg/util/request.go
/*
 * TencentBlueKing is pleased to support the open source community by making
 * 蓝鲸智云 - API 网关(BlueKing - APIGateway) available.
 * Copyright (C) 2025 Tencent. All rights reserved.
 * Licensed under the MIT License (the "License"); you may not use this file except
 * in compliance with the License. You may obtain a copy of the License at
 *
 *     http://opensource.org/licenses/MIT
 *
 * Unless required by applicable law or agreed to in writing, software distributed under
 * the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
 * either express or implied. See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * We undertake not to change the open source license (MIT license) applicable
 * to the current version of the project delivered to anyone in the future.
 */

package util

import (
	"bytes"
	"context"
	"errors"
	"io"
	"net/http"

	"github.com/gin-gonic/gin"

	"mcp_proxy/pkg/constant"
)

// RequestIDKey ...
const RequestIDKey = "request_id"

// InstanceIDKey ...
const InstanceIDKey = "instance_id"

// ErrNilRequestBody ...
var ErrNilRequestBody = errors.New("request Body is nil")

// ReadRequestBody will return the body in []byte, without change the origin body
func ReadRequestBody(r *http.Request) ([]byte, error) {
	if r.Body == nil {
		return nil, ErrNilRequestBody
	}

	body, err := io.ReadAll(r.Body)
	r.Body = io.NopCloser(bytes.NewReader(body))
	return body, err
}

// GetRequestID return the request id from context
func GetRequestID(c *gin.Context) string {
	return c.GetString(RequestIDKey)
}

// GetRequestIDFromContext retrieves the request ID from the context.
func GetRequestIDFromContext(ctx context.Context) string {
	if requestID, ok := ctx.Value(constant.RequestID).(string); ok {
		return requestID
	}
	// Fallback: also check bare string key for backward compatibility with gin.Context.Set
	if requestID, ok := ctx.Value(RequestIDKey).(string); ok {
		return requestID
	}
	return ""
}

// SetRequestID set the request id to context
// nolint:staticcheck
func SetRequestID(c *gin.Context, requestID string) {
	c.Set(RequestIDKey, requestID)
	if c.Request != nil {
		c.Request = c.Request.WithContext(
			context.WithValue(c.Request.Context(), constant.RequestID, requestID))
	}
}

// XRequestIDKey is the context key for the full-chain X-Request-ID
const XRequestIDKey = "x_request_id"

// GetXRequestID return the x_request_id from gin context
func GetXRequestID(c *gin.Context) string {
	return c.GetString(XRequestIDKey)
}

// GetXRequestIDFromContext retrieves the X-Request-ID from the context
func GetXRequestIDFromContext(ctx context.Context) string {
	if xRequestID, ok := ctx.Value(constant.XRequestID).(string); ok {
		return xRequestID
	}
	return ""
}

// SetXRequestID set the x_request_id to context
// nolint:staticcheck
func SetXRequestID(c *gin.Context, xRequestID string) {
	c.Set(XRequestIDKey, xRequestID)
	if c.Request != nil {
		c.Request = c.Request.WithContext(
			context.WithValue(c.Request.Context(), constant.XRequestID, xRequestID))
	}
}

// GetInstanceID get the request id from context
func GetInstanceID(c *gin.Context) string {
	return c.GetString(InstanceIDKey)
}

// SetInstanceID set the request id to context
func SetInstanceID(c *gin.Context, instanceID string) {
	c.Set(InstanceIDKey, instanceID)
}
